DROOLS-5060: [DMN Designer] Editor allows node with prefix/trailing space

manstis opened the pull request:
See https://issues.redhat.com/browse/DROOLS-5060

It also incorporates a fix for https://issues.redhat.com/browse/RHDM-1247

This PR trims leading and trailing space on `Name` values.

It also [normalises whitespace](https://issues.redhat.com/browse/DROOLS-5062?focusedCommentId=13972166&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13972166) in the `Name` value (as requested by @tarilabs).

It's possible to enter the name in the following places:-
- Graph view, node name popover.
- Grid view, name/typeref cell popovers.
- Properties Panel
- Data Types tab (new data-types).
- Imported DMN nodes/data-types `Name` values should remain read-only.

manstis commented:
This PR will be followed up by a change for https://issues.redhat.com/browse/DROOLS-3648

With this change, the _text_ for `InputClause` literal expressions is also trimmed/cleaned.

danielzhe commented:
I'm not sure if this is ok:

![image](https://user-images.githubusercontent.com/7305741/74979888-4315d100-540e-11ea-8f6f-8702c2c87b10.png)

I included a model with spaces in its name.

manstis commented:
Hi @danielzhe this PR only fixes the creation of models with leading/trailing/etc white-spaces in the names. If an imported model contains nodes with white-space in the names they will remain (although I should still not be able to edit the imported name). Was the `     imported        ` prefix the _alias_ given to the import? Perhaps there's an additional place I need to add the check.

danielzhe commented:
@manstis Yes, in the alias. Here:
![image](https://user-images.githubusercontent.com/7305741/75041453-5de25700-549b-11ea-9216-ada6cbfca7c5.png)

manstis commented:
@danielzhe I've fixed naming of imports (in both the "Imports modal" and the "Card view" when renaming an import. This highlighted a bug introduced by this PR that I also fixed).

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/included/grid/BaseCardComponent.java:
```diff
@@ -91,9 +92,9 @@ public HTMLElement getContent() {
+            getIncludedModel().setName(NameUtils.normaliseName(newName));
 
-            if (getIncludedModel().isValid()) {
+            if (getIncludedModel().isValid()) {/**/
```
I think we can remove this /**/

manstis replied:
Fixed in new commit.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/listview/DataTypeListItem.java:
```diff
@@ -430,7 +431,7 @@ private boolean isAReferenceToDestroyedDataType(final DataType dataType) {
         return dataTypeManager
                 .from(dataType)
-                .withName(getName())
+                .withName(NameUtils.normaliseName(getName()))
```
Isn't the jira just about DRG elements? Are data types DRGElements?

manstis replied:
@jomarko The final criteria was `DRGElement` [names](https://issues.redhat.com/browse/DROOLS-5060?focusedCommentId=13972183&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13972183), `ItemDefinition` [names](https://issues.redhat.com/browse/DROOLS-5060?focusedCommentId=13970835&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13970835) and @danielzhe spotted Import [aliases](https://github.com/kiegroup/kie-wb-common/pull/3167#issuecomment-589670573) too.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/property/dmn/NameFieldProvider.java:
```diff
@@ -0,0 +1,61 @@
+@Dependent
+public class NameFieldProvider extends BasicTypeFieldProvider<NameFieldDefinition> {
+
+    //Arbitrary 'magic' number
```
magic number you are saying? ... and who is the magician that do magic with this magic number?

manstis replied:
LOL.. Too much copy and paste from elsewhere.

The priority is used to the Properties Panel code to find a provider for a field definition. Since we only have one provider for `NameFieldDefinition` the value can be anything we like. I've removed the comment and set to `0`.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/included/grid/BaseCardComponentTest.java:
```diff
@@ -144,18 +144,27 @@ public void testGetContent() {
 
         doReturn(true).when(includedModel).isValid();
-        doReturn(emptyList()).when(includedModel).update();
+        doReturn(emptyList()).when(includedModel).update();/**/
```
please remove comment

manstis replied:
Fixed in new commit.

jomarko commented:
### Manual review
Passed, but seems also header cells are affected. I am not saying it is wrong, but have feeling we do not mention it in jira. Am I missing something?

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/property/dmn/NameHolder.java:
```diff
@@ -0,0 +1,80 @@
+
+        final NameHolder name = (NameHolder) o;
+
+        return value != null ? value.equals(name.value) : name.value == null;
```
Fixed in new commit.

manstis commented:
@jomarko In most cases the Header Cell is a container for the containing node's Name (and TypeRef). Therefore the `Name` can be edited in either the _graph_ view or the _grid_ view (via a Header). The only exception I could find was a Decision Table's Input Clause _expression_. This is being addressed by https://issues.redhat.com/browse/DROOLS-3648.

manstis commented:
@danielzhe Please could you re-review when you have a moment.
